---
title: "Linnarsson Data Analysis"
author: "Jonathan Augustin"
date: 
output:
  html_document: default
  pdf_document: default
---

```{r include=FALSE, echo=FALSE}
library(stringr)
library(monocle)
library(Rtsne)
library(reshape2)
library(gapmap)
library(dplyr)
library(magrittr)
library(BiocGenerics)
library(matrixStats)
library(tidyr)
library(GenomicFeatures)
library(RColorBrewer)
library(distr)
library(data.table)
```


```{r Helper_Fxns, include=FALSE, echo=FALSE}

Mode <- function(x) {
  ux <- unique(x)
  ux[which.max(tabulate(match(x, ux)))]
}

```

Generating the CellDataSet object
```{r Data_Import, tidy=TRUE, include=FALSE}
dat<-read.delim('https://storage.googleapis.com/linnarsson-lab-www-blobs/blobs/cortex/expression_mRNA_17-Aug-2014.txt',header=F,stringsAsFactors=F)

# Split out pData 
pData<-as.data.frame(t(dat[1:10,-1]))
colnames(pData)<-as.character(format(pData[1,]))
pData<-pData[-1,]
rownames(pData)<-pData$cell_id

#split out expr data
countData<-dat[-c(1:11),]
rownames(countData)<-str_trim(as.character(format(countData[,1])))
countData<-countData[,-c(1:2)]
colnames(countData)<-pData$cell_id
countData<-data.matrix(countData)

# Split out fData
fData<-data.frame("gene_short_name"=rownames(countData),"clusterAssignment"=countData[,1])

#Get GENCODE Annotation
gencode.dat<- read.table("gencode.vM12.annotation.tab", sep="\t", header = T, stringsAsFactors = F)
rownames(gencode.dat)<-NULL

myJoin<-function(x){paste(x,collapse=",")}

#fData.tmp<-merge(fData,gencode.dat,by.x="gene_short_name",by.y="alias",sort=F,all.x=TRUE)
fData.tmp<-merge(fData, unique(gencode.dat[gencode.dat$ccdsid %in% c("lincRNA", "protein_coding","bidirectional_promoter_lncRNA", "antisense"),c("transcript_name","ccdsid")]), by.x="gene_short_name",by.y="transcript_name",sort=F,all.x=TRUE)

rownames(fData.tmp)<-fData.tmp$gene_short_name
fData.tmp<-fData.tmp[rownames(fData),]
fData<-fData.tmp
fData<- as.data.frame(fData)

# Make CDS object

fd<-new("AnnotatedDataFrame",data=fData)
pd<-new("AnnotatedDataFrame",data=pData)


#New CDS with absolute number of transcripts
dat <-  newCellDataSet(countData, 
                       phenoData = pd, 
                       featureData = fd, 
                       expressionFamily=negbinomial(), 
                       lowerDetectionLimit=1)

dat<-estimateSizeFactors(dat)
dat <- estimateDispersions(dat, cores=2)
dat<-detectGenes(dat,min_expr=1)

# rename 'group #' column
colnames(pData(dat))[2] <- "group_num"

```


```{r Quality_Control, tidy=TRUE, fig.width=15, fig.height=10, echo=F, include=F}
#QC of data after
pData(dat)$Total_mRNAs <- Matrix::colSums(exprs(dat))

qplot(Total_mRNAs, data=pData(dat), fill=level1class, geom="density", alpha=0.1) +
  facet_wrap("level1class", scales = "free") + guides(fill = "none") + theme_bw()

```


```{r Bulk_Analysis, tidy=TRUE, fig.height=5, fig.width=5}
# Rename bidirectionally promoted lncRNAs and antisense lncRNAs to "lincRNAs"
fData(dat)$ccdsid <- gsub("antisense", "lincRNA", fData(dat)$ccdsid)
fData(dat)$ccdsid <- gsub("bidirectional_promoter_lncRNA", "lincRNA", fData(dat)$ccdsid)

# Calculate the mean copies per cell among all classes (Bulk) and draw the density plot for lincRNAs vs protein coding
dat.means<-detectGenes(dat, min_expr = 1)
dat.means<-dat.means[fData(dat.means)$num_cells_expressed >=2]
fData(dat.means)$mean_cpc<-apply(exprs(dat.means),1,mean)
fData(dat.means)$sd<-apply(exprs(dat.means),1,sd)
fData(dat.means)$BCV<-(fData(dat.means)$sd/fData(dat.means)$mean_cpc)^2

tmp<-data.frame("gene_short_name" = fData(dat.means)$gene_short_name,"gene_type" = fData(dat.means)$ccdsid,"mean_cpc"=fData(dat.means)$mean_cpc, "BCV" = fData(dat.means)$BCV, "num_cells_expresed"=fData(dat.means)$num_cells_expressed)

dat_means<-subset(tmp,gene_type %in% c("protein_coding","lincRNA"))

density.plot <- ggplot(dat_means) +
  geom_density(aes(x=log10(mean_cpc),color=gene_type)) + 
  scale_color_manual(values=c("red", "black")) + theme_bw()

density.plot

#List the lincRNAs that are expressed with a mean_cpc greater than 1
dat_lincRNA_sort <- subset(dat_means, gene_type %in% "lincRNA")
dat_mRNA_sort <- subset(dat_means, gene_type %in% "protein_coding")

# Number of lncRNAs
length(dat_lincRNA_sort$gene_short_name)
# Number of mRNAs
length(dat_mRNA_sort$gene_short_name)

#Generate a boxplot on the batch data
dat_means$frac_cells_expressed<-dat_means$num_cells_expresed/length(rownames(pData(dat.means)))

boxplot<-ggplot(dat_means, aes(gene_type, frac_cells_expressed, fill=gene_type)) + geom_boxplot(notch=T) + theme_bw() + scale_fill_manual(values=c("red", "grey50"))

boxplot

#Generate empirical cumulative density function
ecdf<-ggplot(dat_means, aes(x=log10(BCV), color=gene_type)) + stat_ecdf(geom = "line") + theme_bw() + scale_color_manual(values=c("red", "black"))

ecdf

#test dotplot
dot<-ggplot(dat_means, aes(x=log10(mean_cpc), y=frac_cells_expressed, color=gene_type)) + geom_point() + facet_wrap(~gene_type) +theme_bw() + scale_color_manual(values=c("red", "black"))

dot
```


```{r Learned_Distributions, tidy=TRUE, fig.width=5, fig.height=5}
# Density estimates
lnc.dens<-density(subset(dat_means,gene_type %in% "lincRNA")$mean_cpc)
log.lnc.dens<-density(log10(subset(dat_means,gene_type %in% "lincRNA")$mean_cpc))
PC.dens<-density(subset(dat_means,gene_type %in% "protein_coding")$mean_cpc)
log.PC.dens<-density(log10(subset(dat_means,gene_type %in% "protein_coding")$mean_cpc))

#Learn the distribution of lncRNAs
learned.lnc.den<-DiscreteDistribution(subset(dat_means, gene_type %in% "lincRNA")$mean_cpc)

# Create weighted probabilities on PC gene FPKM values from which to sample
PC.weights.on.lnc.D<-p(learned.lnc.den)(subset(dat_means, gene_type %in% "protein_coding")$mean_cpc,lower.tail=FALSE)
PC.probs.on.lnc.D<-PC.weights.on.lnc.D/sum(PC.weights.on.lnc.D)

#Sample from PC genes to match lncRNA distribution
samp_PC<-sample(subset(dat_means, gene_type %in% "protein_coding")$mean_cpc, replace =T, size=339, prob=PC.probs.on.lnc.D)

#My Sampling function
mySample<-function(x,n,EmpDist){
  w<-p(EmpDist)(x$mean_cpc,lower.tail=FALSE)
  probs<-w/sum(w)
  samp<-x[sample(nrow(x),replace=FALSE,size=n,prob=probs),]
  return(samp)
}

#Check plot
plot(density(log10(mySample(subset(dat_means,gene_type %in% "protein_coding"),n=339,EmpDist=learned.lnc.den)$mean_cpc)),main ="Sampling mRNAs from the lncRNA expression distribution")
lines(density(log10(r(learned.lnc.den)(339))), col ="blue")
lines(log.lnc.dens, col ="red")
lines(log.PC.dens,col="green")
legend(x=-0.5,y=0.5,legend=c("lncRNA distribution","Random draws from learned dist","Sampled mRNAs from learned dist"),col=c("red","blue","black"),lty=1) 

#Generate a subset of dat_means that is only mRNAs that fit the learned lncRNA expression distribution
mRNA.sample<-dat_means[dat_means$gene_short_name %in% sample(subset(dat_means, gene_type %in% "protein_coding")$gene_short_name, replace =T, size=339, prob=PC.probs.on.lnc.D),]

#change gene_type to sampled for sampled mRNAs and append it to dat_means
mRNA.sample$gene_type<-"sampled"
dat_sampled<-rbind(dat_means,mRNA.sample)

#Draw a new barplot 
p1<-ggplot(dat_sampled, aes(gene_type, frac_cells_expressed, fill=gene_type)) + geom_boxplot(notch=T) + theme_bw() + scale_fill_manual(values=c("red", "grey50","darkblue"))
p1

#Draw a new ecdf plot
p2<-ggplot(dat_sampled, aes(x=log10(BCV), color=gene_type)) + stat_ecdf(geom = "line") + theme_bw() + scale_color_manual(values=c("red", "black", "darkblue"))
p2

#Draw a new density plot
p3<-ggplot(dat_sampled, aes(x=log10(mean_cpc), color=gene_type)) + geom_density() + theme_bw() + scale_color_manual(values=c("red", "black", "darkblue"))
p3
```

```{r, tidy=T}
#Draw a tSNE using sample mRNAs from above

subsetmRNA<-subset(dat, fData(dat)$gene_short_name %in% mRNA.sample$gene_short_name)

sampled.dat.tSNE<-Rtsne(t(round(vstExprs(subsetmRNA))),theta=0.1, check_duplicates = F) 

pData(subsetmRNA)$tSNE1_pos<-sampled.dat.tSNE$Y[,1]
pData(subsetmRNA)$tSNE2_pos<-sampled.dat.tSNE$Y[,2]

q<-ggplot(pData(subsetmRNA))

sampled_tSNE<-q + geom_point(aes(x=tSNE1_pos,y=tSNE2_pos,color=group_num)) + theme_bw() + coord_equal(1) + scale_color_brewer(palette="Set1")

sampled_tSNE

#There doesn't seem to be any more or less organization present here. May be worth bootstrapping and quantfitying how well each of the tSNEs cluster the data points.
```

```{r tSNE, tidy=TRUE, fig.height=10, fig.width=10}
#tSNE using all genes
dat.tSNE<-Rtsne(t(round(vstExprs(dat))),theta=0.1)

pData(dat)$tSNE1_pos<-dat.tSNE$Y[,1]
pData(dat)$tSNE2_pos<-dat.tSNE$Y[,2]

p<-ggplot(pData(dat))

tSNE_cluster<-p + geom_point(aes(x=tSNE1_pos,y=tSNE2_pos,color=group_num)) + theme_bw() + coord_equal(1) + scale_color_brewer(palette="Set1")

tSNE_cluster

tSNE_level2class<-p + geom_point(aes(x=tSNE1_pos,y=tSNE2_pos,color=level2class)) + theme_bw() + coord_equal(1) +  scale_color_manual(values = colorRampPalette(brewer.pal(9, "Set1"))(length(unique(pData(tmp)$level2class))))

#tSNE using only protein coding genes
pcgenes<-subset(dat, fData(dat)$ccdsid %in% "protein_coding")

protein_coding.dat.tSNE<-Rtsne(t(round(vstExprs(pcgenes))),theta=0.1, check_duplicates = F) 

pData(pcgenes)$tSNE1_pos<-protein_coding.dat.tSNE$Y[,1]
pData(pcgenes)$tSNE2_pos<-protein_coding.dat.tSNE$Y[,2]

q<-ggplot(pData(pcgenes))

protein_coding_tSNE<-q + geom_point(aes(x=tSNE1_pos,y=tSNE2_pos,color=group_num)) + theme_bw() + coord_equal(1) + scale_color_brewer(palette="Set1")

protein_coding_tSNE

#tSNE using only lncRNAs
lncgenes<-subset(dat, fData(dat)$ccdsid %in% "lincRNA")

lincRNA.dat.tSNE<-Rtsne(t(round(vstExprs(lncgenes))),theta=0.1, check_duplicates = F)

pData(lncgenes)$tSNE1_pos<-lincRNA.dat.tSNE$Y[,1]
pData(lncgenes)$tSNE2_pos<-lincRNA.dat.tSNE$Y[,2]

q1<-ggplot(pData(lncgenes))

#tSNE using only lncRNAs color by cluster number (as determined by Linnarsson Lab)
lincRNA_tSNE<-q1 + geom_point(aes(x=tSNE1_pos,y=tSNE2_pos,color=group_num)) + theme_bw() + coord_equal(1) + scale_color_brewer(palette="Set1")

lincRNA_tSNE

#tSNE using only lncRNAs colored by level1class (as determined by Linnarsson Lab)
lincRNA_level1class_tSNE<-q1 + geom_point(aes(x=tSNE1_pos,y=tSNE2_pos,color=level1class)) + theme_bw() + coord_equal(1) + scale_color_brewer(palette="Set1")

lincRNA_level1class_tSNE

#tSNE using only lncRNAs color by level2clss
lincRNA_tSNE_level2class<-q1 + geom_point(aes(x=tSNE1_pos,y=tSNE2_pos,color=level2class)) + theme_bw() + coord_equal(1) +  scale_color_manual(values = colorRampPalette(brewer.pal(9, "Set1"))(length(unique(pData(tmp)$level2class))))

lincRNA_tSNE_level2class

#Color by Sex, size is reflective of XIST expression
lincRNA_tSNE_sex<-q1 + geom_point(aes(x=tSNE1_pos,y=tSNE2_pos,color=sex, size=exprs(lncgenes)[2,])) + scale_size("Xist") + theme_bw() + coord_equal(1) +  scale_fill_brewer(palette = "Set1")

lincRNA_tSNE_sex

#Color by infered batch (utililze the run number from cell_id in order to infer batch)
pData(lncgenes)$cell_id<-colnames(exprs(lncgenes))
pData(lncgenes)$batch<-str_split_fixed(pData(lncgenes)$cell_id, "_", 2)[,1]

j<-ggplot(pData(lncgenes))

lincRNA_tSNE_batch<-j + geom_point(aes(x=tSNE1_pos,y=tSNE2_pos,color=batch)) + theme_bw() + coord_equal(1) + scale_color_manual(values = colorRampPalette(brewer.pal(9, "Set1"))(length(unique(pData(lncgenes)$batch))))

lincRNA_tSNE_batch

#tSNE by lincRNA without Xist and Tsix (we do this because Xist and Tsix expression splits the cells into two groups by sex)
sex_blind.dat<-subset(dat, fData(dat)$ccdsid %in% "lincRNA")
sex_blind_list<-as.vector(rownames(exprs(sex_blind.dat)[c(-2, -6),]))
sex_blind.dat<-subset(sex_blind.dat, rownames(exprs(sex_blind.dat)) %in% sex_blind_list)

sex_blind_lincRNA.tSNE<-Rtsne(t(round(vstExprs(sex_blind.dat))),theta=0.1, check_duplicates = F)

pData(sex_blind.dat)$tSNE1_pos<-sex_blind_lincRNA.tSNE$Y[,1]
pData(sex_blind.dat)$tSNE2_pos<-sex_blind_lincRNA.tSNE$Y[,2]

n<-ggplot(pData(sex_blind.dat))

sex_blind_tSNE<-n + geom_point(aes(x=tSNE1_pos,y=tSNE2_pos,color=group_num)) + theme_bw() + coord_equal(1) + scale_color_brewer(palette="Set1")

sex_blind_tSNE
```

Seperate the "dat" CellDataSet by "Cluster" and calculate mean expression of genes
```{r Cluster, tidy=TRUE, fig.width=25, fig.height=6}
#Seperate the "dat" CellDataSet by "Cluster" and calculate mean expression of genes
Cluster.split<-lapply(unique(pData(dat)$group_num),function(x){ dat[,pData(dat)$group_num==x]})

Cluster.split<-lapply(c(1:length(Cluster.split)), function(x){ detectGenes(Cluster.split[[x]], min_expr = 1)})

Cluster.split<-lapply(c(1:length(Cluster.split)), function(i){
  x<-Cluster.split[[i]]
  x[fData(x)$num_cells_expressed >=2] 
  })

Cluster.split<-lapply(Cluster.split,function(x){
  mean_cpc<-apply(exprs(x),1,mean)
  fData(x)$mean_cpc<-mean_cpc
  return(x)
  })

Cluster.split<-lapply(c(1:length(Cluster.split)), function(i){
  x<-Cluster.split[[i]]
  frac_cells_expressed<-(fData(x)$num_cells_expressed/length(rownames(pData(x))))
  fData(x)$frac_cells_expressed<-frac_cells_expressed
  return(x)
  })

tmp<-data.frame()

group_means<-lapply(c(1:length(Cluster.split)), function(i){
  x<-Cluster.split[[i]]
  res<-data.frame("gene_short_name"=fData(x)$gene_short_name, "gene_type"=fData(x)$ccdsid, "mean_cpc"=fData(x)$mean_cpc, "group_num"=unique(pData(x)$group_num), "frac_cells_expressed"=fData(x)$frac_cells_expressed)
  tmp<-rbind(tmp,res)
  })

tmp<-plyr::ldply(group_means, data.frame)

group_means<-subset(tmp,gene_type %in% c("protein_coding","lincRNA"))

density.plot_Cluster <- ggplot(group_means) +
  geom_density(aes(x=log10(mean_cpc),color=gene_type)) + 
  facet_grid(.~group_num, labeller=labeller(group_num=function(x){paste("Cluster",x,sep=":")})) + 
  scale_color_manual(values=c("red", "black")) + theme_bw()

density.plot_Cluster
```

```{r, tidy=T, fig.height=5, fig.width=5}

#Generate box plots
boxplot_cluster<-ggplot(group_means, aes(gene_type, frac_cells_expressed, fill=gene_type))

boxplot_cluster + geom_boxplot() + theme_bw() + scale_fill_manual(values=c("red", "grey50")) + facet_grid(~ group_num, labeller=labeller(group_num=function(x){paste("Cluster",x,sep=":")}))

#pointplot
group_means1<-data.table(group_means, key="gene_short_name")
group_means1<-group_means1[, .SD[mean_cpc %in% max(mean_cpc)], by=gene_short_name]

group_means2<-data.table(group_means, key="gene_short_name")
group_means2<-group_means2[, .SD[frac_cells_expressed %in% max(frac_cells_expressed) & mean_cpc %in% max(mean_cpc)], by=gene_short_name]

adjusted_point<-ggplot(group_means2, aes(x=log10(mean_cpc), y=frac_cells_expressed, color=gene_type)) + geom_point() + facet_wrap(~gene_type) +theme_bw() + scale_color_manual(values=c("red", "black"))

adjusted_point

```

Seperate the "dat" CellDataSet by level2class and calculate mean expression of genes
```{r Level_2_Class_interneurons, tidy=TRUE,fig.width=25, fig.height=12}
#Seperate the "dat" CellDataSet by the Interneuron level2classes and calculate mean_cpc within this new object
level2.split<-lapply(unique(pData(dat)[pData(dat)$level1class == "interneurons",]$level2class),function(x){
  dat[,pData(dat)$level2class==x]})

level2.split<-lapply(c(1:length(level2.split)), function(x){
  detectGenes(level2.split[[x]], min_expr = 1)})

level2.split<-lapply(c(1:length(level2.split)), function(i){
  x<-level2.split[[i]]
  x[fData(x)$num_cells_expressed >=2] 
  })

level2.split<-lapply(level2.split,function(x){
  mean_cpc<-apply(exprs(x),1,mean)
  fData(x)$mean_cpc<-mean_cpc
  return(x)
})

tmp<-data.frame()

group_means_level2class<-lapply(c(1:length(level2.split)), function(i){
  x<-level2.split[[i]]
  res<-data.frame("gene_short_name"=fData(x)$gene_short_name ,"gene_type"=fData(x)$ccdsid ,"mean_cpc"=fData(x)$mean_cpc ,"level2class"=unique(pData(x)$level2class))
  tmp<-rbind(tmp,res)
  })

tmp<-plyr::ldply(group_means_level2class, data.frame)

group_means_level2class<-subset(tmp,gene_type %in% c("protein_coding","lincRNA"))

density.plot_level2class <- ggplot(group_means_level2class) +
  geom_density(aes(x=log10(mean_cpc),color=gene_type)) + 
  facet_wrap(~level2class, nrow = 2) + 
  scale_color_manual(values=c("red", "black")) + theme_bw()

density.plot_level2class
```

```{r, tidy=TRUE,fig.width=5, fig.height=5}
#Generate a barplot that has all of the subcell types for each of the 9 clusters (again as determined by the Linnarsson group)
level_2_barplot<-lapply(c(1:9), function(x){ pData(Cluster.split[[x]])$level2class %>% unique() %>% length()})
level_2_barplot<-data.frame(level_2_barplot)
colnames(level_2_barplot)<- c("1", "2", "3", "4", "5", "6", "7", "8", "9")
rownames(level_2_barplot)<-"Number_of_Clusters"
level_2_barplot<-as.data.frame(level_2_barplot)

level_2_barplot<-as.data.frame(t(level_2_barplot))

bar.plot <- ggplot(level_2_barplot)

bar.plot + geom_bar(aes(x=rownames(level_2_barplot), y=level_2_barplot$Number_of_Clusters, fill=rownames(level_2_barplot)), colour="black", width=0.5, stat = "identity") + scale_fill_brewer(palette = "Set1", guide = guide_legend(title = "Cluster #")) + xlab("Cluster #") + ylab("Number of cellular subtypes") + theme_bw() + theme(axis.text.x = element_blank())

```

```{r Level2class_complete_split, tidy=TRUE, fig.height=25, fig.width=25}
#Completely split the CellDataSet by level2class and calculate the mean_cpc in each of the new CellDataSet objects
complete.level2.split<-lapply(unique(pData(dat)$level2class),function(x){
  dat[,pData(dat)$level2class==x]})

complete.level2.split<-lapply(c(1:length(unique(pData(dat)$level2class))), function(x){
  detectGenes(complete.level2.split[[x]], min_expr = 1)})

complete.level2.split<-lapply(c(1:length(unique(pData(dat)$level2class))), function(i){
  x<-complete.level2.split[[i]]
  x[fData(x)$num_cells_expressed >=2] 
  })

complete.level2.split<-lapply(complete.level2.split,function(x){
  mean_cpc<-apply(exprs(x),1,mean)
  fData(x)$mean_cpc<-mean_cpc
  return(x)
  })

complete.level2.split<- lapply(c(1:length(unique(pData(dat)$level2class))), function(i){
  x=complete.level2.split[[i]]
  fData(x)$gene_std_pc<-rowSds(exprs(x))
  return(x)
  })

complete.level2.split<-lapply(c(1:length(unique(pData(dat)$level2class))), function(i){
  x=complete.level2.split[[i]]
  fData(x)$BCV<-as.vector(fData(x)$gene_std_pc)/as.vector(fData(x)$mean_cpc)
  return(x)
  })

complete.level2.split<-lapply(c(1:length(unique(pData(dat)$level2class))), function(i){
  x=complete.level2.split[[i]]
  fData(x)$frac_cells_expressed<-fData(x)$num_cells_expressed/length(rownames(pData(x)))
  return(x)
  })
  
tmp<- data.frame()

bcv_level2class<-lapply(c(1:length(complete.level2.split)), function(i){
  x<-complete.level2.split[[i]]
  res<-data.frame("gene_short_name"=fData(x)$gene_short_name, "gene_type"=fData(x)$ccdsid, "mean_cpc"=fData(x)$mean_cpc,"BCV" =fData(x)$BCV, "gene_sd_pc" = fData(x)$gene_std_pc, "level2class"=unique(pData(x)$level2class), "frac_cells_expressed"=fData(x)$frac_cells_expressed)
  tmp<-rbind(tmp,res)
  })

names(bcv_level2class)<-unique(pData(dat)$level2class)

level2_bcv<-Reduce(function(...) merge(..., all=TRUE), bcv_level2class)
level2_bcv<-subset(level2_bcv, gene_type %in% c("lincRNA","protein_coding"))

#Generate density plots
density.plot_all_level2class <- ggplot(level2_bcv) +
  geom_density(aes(x=log10(mean_cpc),color=gene_type)) + 
  facet_wrap(~level2class, nrow = 8) + 
  scale_color_manual(values=c("red", "black")) + theme_bw()

density.plot_all_level2class

#Generate box plots
boxplot_level2Class<-ggplot(level2_bcv, aes(gene_type, frac_cells_expressed, fill=gene_type))
boxplot_level2Class + geom_boxplot() + theme_bw() + scale_fill_manual(values=c("red", "grey50")) + facet_wrap(~ level2class, nrow=8)

#generate point plots
level2_bcv1<-data.table(level2_bcv, key = "gene_short_name")
level2_bcv1<-level2_bcv1[, .SD[frac_cells_expressed %in% max(frac_cells_expressed) & mean_cpc %in% max(mean_cpc)], by=gene_short_name]

level2class_adjusted_point<-ggplot(level2_bcv1, aes(x=log10(mean_cpc), y=frac_cells_expressed, color=gene_type)) + geom_point() + facet_wrap(~gene_type) +theme_bw() + scale_color_manual(values=c("red", "black"))

level2class_adjusted_point

#sanity test density plot
q1<-ggplot(level2_bcv1, aes(x=log10(mean_cpc), color=gene_type))+geom_density()+theme_bw()+scale_color_manual(values=c("red", "black"))

#summary of level2class analysis

```

```{r Calculate_STD_&_BCV, tidy=TRUE, fig.width=5, fig.height=5}
#Some odds and ends that I may want to come back and use later.
cluster.bcv.dat<-lapply(unique(pData(dat)$group_num),function(x){ dat[,pData(dat)$group_num==x]})

cluster.bcv.dat<-lapply(c(1:length(cluster.bcv.dat)), function(x){ detectGenes(cluster.bcv.dat[[x]], min_expr = 1)})

cluster.bcv.dat<-lapply(c(1:length(cluster.bcv.dat)), function(i){
  x<-cluster.bcv.dat[[i]]
  x[fData(x)$num_cells_expressed >=2] 
  })

cluster.bcv.dat<-lapply(cluster.bcv.dat, function(x){
  mean_cpc<-apply(exprs(x), 1, mean)
  fData(x)$mean_cpc<-mean_cpc
  return(x)
  })

cluster.bcv.dat<- lapply(c(1:length(cluster.bcv.dat)), function(i){
  x=cluster.bcv.dat[[i]]
  fData(x)$gene_std_pc<-rowSds(exprs(x))
  return(x)
  })

cluster.bcv.dat<-lapply(c(1:length(unique(cluster.bcv.dat))), function(i){
  x=cluster.bcv.dat[[i]]
  fData(x)$BCV<-as.vector(fData(x)$gene_std_pc)/as.vector(fData(x)$mean_cpc)
  return(x)
  })

tmp<- data.frame()

bcv_cluster<-lapply(c(1:length(cluster.bcv.dat)), function(i){
  x<-cluster.bcv.dat[[i]]
  res<-data.frame("gene_short_name"=fData(x)$gene_short_name, "gene_type"=fData(x)$ccdsid, "mean_cpc"=fData(x)$mean_cpc,"BCV" =fData(x)$BCV, "gene_sd_pc" = fData(x)$gene_std_pc, "cluster_num" = rep(i))
  tmp<-rbind(tmp,res)
  })

bcv<-Reduce(function(...) merge(..., all=TRUE), bcv_cluster)

linc_bcv<-subset(bcv, bcv$gene_type=="lincRNA")
mRNA_bcv<-subset(bcv, bcv$gene_type=="protein_coding")

delta_median_by_cluster<-lapply(c(1:9), function(x){
  median(subset(mRNA_bcv, mRNA_bcv$cluster_num==x)$mean_cpc)-median(subset(linc_bcv,linc_bcv$cluster_num==x)$mean_cpc)
  })

names(delta_median_by_cluster)<-c("Cluster 1", "Cluster 2", "Cluster 3", "Cluster 4", "Cluster 5", "Cluster 6", "Cluster 7", "Cluster 8", "Cluster 9")

#Generate a scatter plot of Δ median copies per cell (mRNA-lncRNA) for each cluster
delta_median_by_cluster<-as.matrix(delta_median_by_cluster)
delta_median_by_cluster<-as.data.frame(delta_median_by_cluster)
colnames(delta_median_by_cluster) <- "delta_median_mean_cpc"
delta_median_by_cluster$delta_median_mean_cpc<-as.numeric(delta_median_by_cluster$delta_median_mean_cpc)

delta_median_by_cluster$Number_of_subtypes<- as.vector(level_2_barplot$Number_of_Clusters)

p<-ggplot(delta_median_by_cluster, aes(x=Number_of_subtypes, y=delta_median_mean_cpc))

scatter.plot<-p + geom_point(aes(color=rownames(delta_median_by_cluster))) + geom_smooth(aes(alpha=0.1),method = "lm", se = T, color="black", alpha=0.2) + scale_color_brewer(palette = "Set1", guide = guide_legend(title = "Cluster #")) + guides(fill="none") + theme_bw() + xlab("Number of cellular subtypes") +ylab("Δ median copies per cell (mRNA-lncRNA)") + theme(legend.position="none")

scatter.plot

#generate ecdf plots
subset_bcv<-subset(bcv, bcv$gene_type %in% c("lincRNA","protein_coding"))
ecdf_cluster<-ggplot(subset_bcv, aes(x=log10(BCV), color=gene_type)) + stat_ecdf(geom = "line") + facet_wrap(~cluster_num, nrow=1) + scale_color_manual(values=c("red", "black")) + theme_bw()

ecdf_cluster
```
#Session Info
```{r, Session Info, tidy=T}
sessionInfo()

