---
title: "Linnarsson Data Analysis"
author: "Jonathan Augustin"
date: "1/9/2017"
output: pdf_document
---

```{r include=FALSE, echo=FALSE}
library(stringr)
library(monocle)
library(Rtsne)
library(reshape2)
library(gapmap)
library(dplyr)
library(magrittr)
library(BiocGenerics)
library(matrixStats)
library(tidyr)
library(GenomicFeatures)
```

Generating the CellDataSet object
```{r Data_Import, tidy=TRUE, include=FALSE}
dat<-read.delim('https://storage.googleapis.com/linnarsson-lab-www-blobs/blobs/cortex/expression_mRNA_17-Aug-2014.txt',header=F,stringsAsFactors=F)

# Split out pData 
pData<-as.data.frame(t(dat[1:10,-1]))
colnames(pData)<-as.character(format(pData[1,]))
pData<-pData[-1,]
rownames(pData)<-pData$cell_id

#split out expr data
countData<-dat[-c(1:11),]
rownames(countData)<-str_trim(as.character(format(countData[,1])))
countData<-countData[,-c(1:2)] ### Need to check with Loyal about this... (nevermind...)
colnames(countData)<-pData$cell_id
countData<-data.matrix(format(countData))

# Split out fData
fData<-data.frame("gene_short_name"=rownames(countData),"clusterAssignment"=countData[,1])

#Get GENCODE Annotation
gencode.dat<- read.table("gencode.vM12.annotation.tab", sep="\t", header = T, stringsAsFactors = F)
rownames(gencode.dat)<-NULL

myJoin<-function(x){paste(x,collapse=",")}

#fData.tmp<-merge(fData,ucsc.dat,by.x="gene_short_name",by.y="alias",sort=F,all.x=TRUE)
fData.tmp<-merge(fData, unique(gencode.dat[gencode.dat$ccdsid %in% c("lincRNA", "protein_coding"),c("transcript_name","ccdsid")]), by.x="gene_short_name",by.y="transcript_name",sort=F,all.x=TRUE)
as.data.frame(fData.tmp)
rownames(fData.tmp)<-fData.tmp$gene_short_name
fData<-fData.tmp[rownames(fData),]
colnames(fData)<-c("gene_short_name", "clusterAssignment", "transcript_type")
rownames(fData.tmp)<-fData.tmp$gene_short_name

# Make CDS object

fd<-new("AnnotatedDataFrame",data=fData)
pd<-new("AnnotatedDataFrame",data=pData)


#New CDS with absolute number of transcripts
dat <-  newCellDataSet(countData, 
                       phenoData = pd, 
                       featureData = fd, 
                       expressionFamily=negbinomial(), 
                       lowerDetectionLimit=1)

# rename 'group #' column
colnames(pData(dat))[2] <- "group_num"

```


```{r Quality_Control, tidy=TRUE, fig.width=15, fig.height=10 }
#QC of data after
pData(dat)$Total_mRNAs <- Matrix::colSums(exprs(dat))

qplot(Total_mRNAs, data=pData(dat), fill=level1class, geom="density", alpha=0.1) +
  facet_wrap("level1class", scales = "free") + guides(fill = "none") + theme_bw()

```


```{r Bulk_Analysis, tidy=TRUE, fig.height=5, fig.width=5}
# Calculate the mean copies per cell among all classes (Bulk) and draw the density plot for lincRNAs vs protein coding
dat.means<-detectGenes(dat, min_expr = 1)
dat.means<-dat.means[fData(dat.means)$num_cells_expressed >= 1]
fData(dat.means)$mean_cpc<-apply(exprs(dat.means),1,mean)

tmp<-data.frame("gene_short_name" = fData(dat.means)$gene_short_name,"gene_type" = fData(dat.means)$transcript_type,"mean_cpc"=fData(dat.means)$mean_cpc)

dat_means<-subset(tmp,gene_type %in% c("protein_coding","lincRNA"))

density.plot <- ggplot(dat_means) +
  geom_density(aes(x=log10(mean_cpc),color=gene_type)) + 
  scale_color_manual(values=c("red", "black")) + theme_bw()

density.plot

#List the lincRNAs that are expressed with a mean_cpc greater than 1
dat_lincRNA_sort <- subset(dat_means, gene_type %in% "lincRNA")
dat_mRNA_sort <- subset(dat_means, gene_type %in% "protein_coding")

#length(dat_lincRNA_sort$gene_short_name)
print("Number of lncRNAs = 441")
#length(dat_mRNA_sort$gene_short_name)
print("Number of mRNAs = 17091")

```

Seperate the "dat" CellDataSet by "Cluster" and calculate mean expression of genes
```{r Cluster, tidy=TRUE, fig.width=25, fig.height=6}
#Seperate the "dat" CellDataSet by "Cluster" and calculate mean expression of genes
Cluster.split<-lapply(unique(pData(dat)$group_num),function(x){ dat[,pData(dat)$group_num==x]})

Cluster.split<-lapply(c(1:length(Cluster.split)), function(x){ detectGenes(Cluster.split[[x]], min_expr = 1)})

Cluster.split<-lapply(c(1:length(Cluster.split)), function(i){
  x<-Cluster.split[[i]]
  x[fData(x)$num_cells_expressed > 1] 
  })

Cluster.split<-lapply(Cluster.split,function(x){
  mean_cpc<-apply(exprs(x),1,mean)
  fData(x)$mean_cpc<-mean_cpc
  return(x)
})

tmp<-data.frame()
group_means<-lapply(c(1:length(Cluster.split)), function(i){
  x<-Cluster.split[[i]]
  res<-data.frame("gene_short_name"=fData(x)$gene_short_name, "gene_type"=fData(x)$transcript_type, "mean_cpc"=fData(x)$mean_cpc, "group_num"=unique(pData(x)$group_num))
  tmp<-rbind(tmp,res)
  })

tmp<-plyr::ldply(group_means, data.frame)

group_means<-subset(tmp,gene_type %in% c("protein_coding","lincRNA"))

density.plot_Cluster <- ggplot(group_means) +
  geom_density(aes(x=log10(mean_cpc),color=gene_type)) + 
  facet_grid(.~group_num, labeller=labeller(group_num=function(x){paste("Cluster",x,sep=":")})) + 
  scale_color_manual(values=c("red", "black")) + theme_bw()

density.plot_Cluster
```

Seperate the "dat" CellDataSet by level2class and calculate mean expression of genes
```{r Level_2_Class_interneurons, tidy=TRUE,fig.width=25, fig.height=12}
#Seperate the "dat" CellDataSet by level2class and calculate mean expression of genes
level2.split<-lapply(unique(pData(dat)[pData(dat)$level1class == "interneurons",]$level2class),function(x){
  dat[,pData(dat)$level2class==x]})

level2.split<-lapply(c(1:length(level2.split)), function(x){
  detectGenes(level2.split[[x]], min_expr = 1)})

level1.split<-lapply(c(1:length(level2.split)), function(i){
  x<-level2.split[[i]]
  x[fData(x)$num_cells_expressed >=3] 
  })

level2.split<-lapply(level2.split,function(x){
  mean_cpc<-apply(exprs(x),1,mean)
  fData(x)$mean_cpc<-mean_cpc
  return(x)
})

tmp<-data.frame()

group_means_level2class<-lapply(c(1:length(level2.split)), function(i){
  x<-level2.split[[i]]
  res<-data.frame("gene_short_name"=fData(x)$gene_short_name ,"gene_type"=fData(x)$transcript_type ,"mean_cpc"=fData(x)$mean_cpc ,"level2class"=unique(pData(x)$level2class))
  tmp<-rbind(tmp,res)
  })

tmp<-plyr::ldply(group_means_level2class, data.frame)

group_means_level2class<-subset(tmp,gene_type %in% c("protein_coding","lincRNA"))

density.plot_level2class <- ggplot(group_means_level2class) +
  geom_density(aes(x=log10(mean_cpc),color=gene_type)) + 
  facet_wrap(~level2class, nrow = 2) + 
  scale_color_manual(values=c("red", "black")) + theme_bw()

density.plot_level2class
```

```{r, tidy=TRUE,fig.width=5, fig.height=5}

level_2_barplot<-lapply(c(1:9), function(x){ pData(Cluster.split[[x]])$level2class %>% unique() %>% length()})
level_2_barplot<-data.frame(level_2_barplot)
colnames(level_2_barplot)<- c("Cluster_1", "Cluster_2", "Cluster_3", "Cluster_4", "Cluster_5", "Cluster_6", "Cluster_7", "Cluster_8", "Cluster_9")
rownames(level_2_barplot)<-"Number_of_Clusters"
level_2_barplot<-as.data.frame(level_2_barplot)
as.matrix(level_2_barplot)

tmp<-as.data.frame(t(level_2_barplot))

bar.plot <- ggplot(tmp)

bar.plot + geom_bar(aes(x=rownames(tmp), y=tmp$Number_of_Clusters, fill=rownames(tmp)), colour="black", width=0.5, stat = "identity") + scale_fill_brewer(palette = "Set1") + xlab("Cluster") + ylab("Number of cellular subtypes") + guides(fill="none") + theme_bw()
```

```{r Calculate_STD_&_BCV, tide}

level2.split<- lapply(c(1:length(level2.split)), function(i){
  x=level2.split[[i]]
  fData(x)$gene_std_pc<-rowSds(exprs(x))
  return(x)
})

level2.split<-lapply(c(1:length(level2.split)), function(i){
  x=level2.split[[i]]
  fData(x)$BCV<-as.vector(fData(x)$gene_std_pc)/as.vector(fData(x)$mean_cpc)
  return(x)
})

tmp<- data.frame()

bcv_cluster<-lapply(c(1:length(level2.split)), function(i){
  x<-level2.split[[i]]
  res<-data.frame("gene_short_name"=fData(x)$gene_short_name, "gene_type"=fData(x)$transcript_type, "mean_cpc"=fData(x)$mean_cpc,"BCV" =fData(x)$BCV, "gene_sd_pc" = fData(x)$gene_std_pc, "level2class"=unique(pData(x)$level2class))
  tmp<-rbind(tmp,res)
  })
  
 scatter.plot <- ggplot(bcv_cluster) + geom_point(aes(x=log10(BCV), 
                                                             y=level2class, 
                                                             color=)) + 
  theme_bw() + facet_grid(. ~ gene_type) + scale_color_manual(values=c("red", "black"))
scatter.plot
```

```{r}

complete.level2.split<-lapply(unique(pData(dat)$level2class),function(x){
  dat[,pData(dat)$level2class==x]})

complete.level2.split<-lapply(c(1:length(complete.level2.split)), function(x){
  detectGenes(complete.level2.split[[x]], min_expr = 1)})

complete.level1.split<-lapply(c(1:length(complete.level2.split)), function(i){
  x<-complete.level2.split[[i]]
  x[fData(x)$num_cells_expressed >=3] 
  })

complete.level2.split<-lapply(complete.level2.split,function(x){
  mean_cpc<-apply(exprs(x),1,mean)
  fData(x)$mean_cpc<-mean_cpc
  return(x)
  })


```



